File: packages/sdk/src/crdt/presence.ts
import { Schema } from "effect";
import { encode } from "../codec.js";
import type { WsTransport } from "../transport/websocket.js";
import type { PresenceDelta, PresenceEntryDelta } from "../sync/delta.js";

export interface PresenceEntry<T> {
  clientId: number;
  data: T;
  expiresAtMs: number;
}

/**
 * Low-level handle for a presence channel CRDT.
 *
 * Obtained via `MeridianClient.presence()`. Prefer the `usePresence` React hook
 * for component-level usage (it manages heartbeat timers and cleanup automatically);
 * use this handle directly in non-React environments.
 */
export class PresenceHandle<T> {
  private readonly entries = new Map<string, PresenceEntry<T>>();
  private readonly crdtId: string;
  private readonly clientId: number;
  private readonly transport: WsTransport;
  private readonly schema: Schema.Schema<T> | null;
  private readonly listeners = new Set<(entries: PresenceEntry<T>[]) => void>();

  constructor(opts: {
    ns: string;
    crdtId: string;
    clientId: number;
    transport: WsTransport;
    schema?: Schema.Schema<T>;
  }) {
    this.crdtId = opts.crdtId;
    this.clientId = opts.clientId;
    this.transport = opts.transport;
    this.schema = opts.schema ?? null;
  }

  /**
   * Returns all currently online entries — those whose TTL has not yet elapsed
   * as of the call time.
   */
  online(): PresenceEntry<T>[] {
    const now = Date.now();
    const live: PresenceEntry<T>[] = [];
    for (const entry of this.entries.values()) {
      if (entry.expiresAtMs > now) live.push(entry);
    }
    return live;
  }

  /**
   * Registers a listener that is called whenever the online entries change.
   *
   * @returns An unsubscribe function — call it to stop receiving updates.
   */
  onChange(listener: (entries: PresenceEntry<T>[]) => void): () => void {
    this.listeners.add(listener);
    return () => { this.listeners.delete(listener); };
  }

  /**
   * Broadcasts a heartbeat that marks the current client as online for `ttlMs`
   * milliseconds (default `30 000`).
   *
   * Call this periodically (at least once per `ttlMs`) to remain visible to
   * other clients. The `usePresence` hook manages this automatically when
   * `opts.data` is provided.
   *
   * @param data - Arbitrary payload broadcast to other clients.
   * @param ttlMs - Time-to-live in milliseconds before the entry is considered stale.
   */
  heartbeat(data: T, ttlMs: number = 30_000): void {
    const wallMs = Date.now();
    const hlc = { wall_ms: wallMs, logical: 0, node_id: this.clientId };

    this.entries.set(String(this.clientId), {
      clientId: this.clientId,
      data,
      expiresAtMs: wallMs + ttlMs,
    });
    this.emit();

    this.transport.send({
      Op: {
        crdt_id: this.crdtId,
        op_bytes: encode({
          Presence: { Heartbeat: { client_id: this.clientId, data, hlc, ttl_ms: ttlMs } },
        }),
      },
    });
  }

  /**
   * Marks the current client as offline and broadcasts a leave operation.
   *
   * After calling this, the client's entry is removed from the `online()` list
   * on all peers. The `usePresence` hook calls this automatically on unmount.
   */
  leave(): void {
    const wallMs = Date.now();
    const hlc = { wall_ms: wallMs, logical: 0, node_id: this.clientId };

    this.entries.delete(String(this.clientId));
    this.emit();

    this.transport.send({
      Op: {
        crdt_id: this.crdtId,
        op_bytes: encode({ Presence: { Leave: { client_id: this.clientId, hlc } } }),
      },
    });
  }

  applyDelta(delta: PresenceDelta): void {
    let changed = false;
    for (const [clientIdStr, entry] of Object.entries(delta.changes)) {
      if (entry === null) {
        if (this.entries.has(clientIdStr)) {
          this.entries.delete(clientIdStr);
          changed = true;
        }
      } else {
        changed = this.mergeEntry(clientIdStr, entry) || changed;
      }
    }
    if (changed) this.emit();
  }

  private decode(raw: unknown): T {
    if (this.schema !== null) {
      return Schema.decodeUnknownSync(this.schema)(raw);
    }
    return raw as T;
  }

  private mergeEntry(clientIdStr: string, incoming: PresenceEntryDelta): boolean {
    const existing = this.entries.get(clientIdStr);
    const incomingExpiresAt = Number(incoming.hlc.wall_ms) + Number(incoming.ttl_ms);

    if (existing === undefined) {
      if (incoming.ttl_ms > 0) {
        this.entries.set(clientIdStr, {
          clientId: Number(clientIdStr),
          data: this.decode(incoming.data),
          expiresAtMs: incomingExpiresAt,
        });
        return true;
      }
      return false;
    }

    if (incomingExpiresAt >= existing.expiresAtMs) {
      if (incoming.ttl_ms === 0) {
        this.entries.delete(clientIdStr);
      } else {
        this.entries.set(clientIdStr, {
          clientId: Number(clientIdStr),
          data: this.decode(incoming.data),
          expiresAtMs: incomingExpiresAt,
        });
      }
      return true;
    }
    return false;
  }

  private emit(): void {
    const live = this.online();
    for (const listener of this.listeners) listener(live);
  }
}

File: packages/python/src/meridian/crdt/_presence.py
import asyncio
import time
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from .._codec import encode

if TYPE_CHECKING:
    from .._transport import Transport


@dataclass
class PresenceEntry:
    client_id: int
    data: Any
    expires_at_ms: int


class Presence:
    """Presence CRDT — tracks which clients are currently online with optional metadata.

    Each client broadcasts a heartbeat to signal that it's still alive.
    Entries expire automatically based on ``ttl_ms``.
    """

    def __init__(
        self,
        crdt_id: str,
        client_id: int,
        transport: "Transport",
        *,
        encrypt_fn=None,
        decrypt_fn=None,
    ) -> None:
        self._id = crdt_id
        self._client_id = client_id
        self._transport = transport
        self._encrypt_fn = encrypt_fn
        self._decrypt_fn = decrypt_fn
        self._entries: dict[str, PresenceEntry] = {}
        self._listeners: list[asyncio.Queue] = []

    # ── read ─────────────────────────────────────────────────────────────────

    def online(self) -> list[PresenceEntry]:
        now = int(time.time() * 1000)
        return [e for e in self._entries.values() if e.expires_at_ms > now]

    # ── write ─────────────────────────────────────────────────────────────────

    def join(self, data: Any, *, ttl_ms: int = 30_000) -> None:
        t = asyncio.create_task(self._send_heartbeat(data, ttl_ms))
        t.add_done_callback(lambda _: None)

    async def join_async(self, data: Any, *, ttl_ms: int = 30_000) -> None:
        await self._send_heartbeat(data, ttl_ms)

    def leave(self) -> None:
        wall_ms = int(time.time() * 1000)
        hlc = {"wall_ms": wall_ms, "logical": 0, "node_id": self._client_id}
        op = encode({"Presence": {"Leave": {"client_id": self._client_id, "hlc": hlc}}})
        self._transport.send({"Op": {"crdt_id": self._id, "op_bytes": op}})
        self._entries.pop(str(self._client_id), None)
        self._notify()

    # ── subscribe ─────────────────────────────────────────────────────────────

    async def changes(self) -> AsyncIterator[list[PresenceEntry]]:
        q: asyncio.Queue = asyncio.Queue()
        self._listeners.append(q)
        try:
            while True:
                yield await q.get()
        finally:
            self._listeners.remove(q)

    # ── internal ──────────────────────────────────────────────────────────────

    async def _send_heartbeat(self, data: Any, ttl_ms: int) -> None:
        wall_ms = int(time.time() * 1000)
        hlc = {"wall_ms": wall_ms, "logical": 0, "node_id": self._client_id}
        wire_data = await self._encrypt_fn(data) if self._encrypt_fn else data
        op = encode(
            {
                "Presence": {
                    "Heartbeat": {
                        "client_id": self._client_id,
                        "data": wire_data,
                        "hlc": hlc,
                        "ttl_ms": ttl_ms,
                    }
                }
            }
        )
        self._transport.send({"Op": {"crdt_id": self._id, "op_bytes": op}})

    async def _apply_delta(self, delta: dict) -> None:
        changed = False
        for client_id_str, entry in (delta.get("changes") or {}).items():
            if entry is None:
                if client_id_str in self._entries:
                    del self._entries[client_id_str]
                    changed = True
            else:
                hlc = entry.get("hlc", {})
                ttl = int(entry.get("ttl_ms", 0))
                wall = int(hlc.get("wall_ms", 0))
                expires = wall + ttl
                data = entry.get("data")
                if self._decrypt_fn and isinstance(data, dict) and data.get("$e") == 1:
                    data = await self._decrypt_fn(data)
                new_entry = PresenceEntry(
                    client_id=int(client_id_str),
                    data=data,
                    expires_at_ms=expires,
                )
                existing = self._entries.get(client_id_str)
                if existing is None or expires > existing.expires_at_ms:
                    self._entries[client_id_str] = new_entry
                    changed = True
        if changed:
            self._notify()

    def _notify(self) -> None:
        entries = self.online()
        for q in self._listeners:
            try:
                q.put_nowait(entries)
            except asyncio.QueueFull:
                pass
